### PR TITLE
Hsjfelix refactor

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by ghost** [comment]
The refactored Request, Response and StaticFileHandler classes work on their own and are unit tested. I doesn't as a whole since I haven't refactor other codes accordingly. Some files in the server folder should be deleted because of the refactoring

**Comment by ghost** [comment]
Remeber to remove the Nignx namespace

File: server/file_handler.cc [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+#include <string.h>
+#include <cstdio>
+#include "http.h"
+
+RequestHandler::Status StaticFileHandler::Init(const std::string& uri_prefix, const NginxConfig& config){
+	this->m_uri_prefix_ = uri_prefix;
+	//assume the config argument has something like
+	// root /path1
+	//check if such a line exists
+	auto root_config_tokens = config.find("root");
```
Comment (richardmin) [inline server/file_handler.cc]: I'm not a big fan of using auto here. Automatically inferring types isn't a good practice

Comment (richardmin) [reply server/file_handler.cc]: I believe it leads to code obscurity in the future.

File: server/file_handler.cc [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	if(root_config_tokens.size() < 2){
+		printf("StaticfileHandler.Init: Invalid config");
+		return INVALID_CONFIG;
+	}
```
Comment (richardmin) [inline server/file_handler.cc]: In general, if's should have whitespace
i.e. this should be 
`if (root_config_tokens.size() < 2) {`

File: server/file_handler.cc [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+#include "http.h"
+
+RequestHandler::Status StaticFileHandler::Init(const std::string& uri_prefix, const NginxConfig& config){
+	this->m_uri_prefix_ = uri_prefix;
+	//assume the config argument has something like
+	// root /path1
+	//check if such a line exists
+	auto root_config_tokens = config.find("root");
+	if(root_config_tokens.size() < 2){
+		printf("StaticfileHandler.Init: Invalid config");
```
Comment (richardmin) [inline server/file_handler.cc]: We should also print which block is invalid (i.e. the URI prefix)

File: server/file_handler.cc [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+		return INVALID_CONFIG;
+	}
+	this->m_root_path_ = root_config_tokens[1];
+	return OK;
+}
+
+RequestHandler::Status StaticFileHandler::HandleRequest(const Request& request, Response* response){
+	std::string req_uri = request.uri();
+	std::size_t prefix_pos = req_uri.find(this->m_uri_prefix_);
+	if(prefix_pos == std::string::npos || prefix_pos != 0){
```
Comment (richardmin) [inline server/file_handler.cc]: Again, the if whitespacing.

File: server/file_handler.cc [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+	auto root_config_tokens = config.find("root");
+	if(root_config_tokens.size() < 2){
+		printf("StaticfileHandler.Init: Invalid config");
+		return INVALID_CONFIG;
+	}
+	this->m_root_path_ = root_config_tokens[1];
+	return OK;
+}
+
+RequestHandler::Status StaticFileHandler::HandleRequest(const Request& request, Response* response){
```
Comment (richardmin) [inline server/file_handler.cc:21]: Whitespacing for `...response) {`

File: server/file_handler.cc [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	std::string req_uri = request.uri();
```
Comment (richardmin) [inline server/file_handler.cc:22]: We should use two spaces instead of tabs for indentation

File: server/file_handler.cc [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+	std::size_t prefix_pos = req_uri.find(this->m_uri_prefix_);
+	if(prefix_pos == std::string::npos || prefix_pos != 0){
+		printf("StaticfileHandler.HandleRequest: Bad request");
+		response->SetStatus(Response::BAD_REQUEST);
+		response->AddHeader("Content-Type",http::mime_type::ContentTypeAsString(http::mime_type::CONTENT_TYPE_TEXT_HTML));
+		return BAD_REQUEST;
+	}
+
```
Comment (richardmin) [inline server/file_handler.cc:32]: I believe we don't actually have to validate the URI -- instead, the webserver should pass the appropriate URI to the appropriate handler. This means that this (type) of code should reside in the webserver.

File: server/file_handler.cc [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+	std::string uri_no_prefix = req_uri.substr(this->m_uri_prefix_.size());
+	std::string actual_uri = this->m_root_path_ + uri_no_prefix;
+
+	if(!FileIO::FileExists(actual_uri)){
+		printf("StaticfileHandler.HandleRequest: File not found:");
+		printf("%s\n", actual_uri.c_str());
```
Comment (richardmin) [inline server/file_handler.cc]: These two lines can be reduced to
`printf("StaticFileHandler.HandleRequest: File Not Found %s\n", actual_uri.c_str());`

File: server/file_handler.cc [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    }
+    else {
+      std::string file_extension = actual_uri.substr(index_of_extension + 1);
+      response->AddHeader("Content-Type",http::mime_type::ContentTypeAsString(http::mime_type::GetMimeType(file_extension)));
+    }
+    
+    std::string body;
+    FileIO::FileToString(actual_uri, body);
+    response->SetBody(body);
+
```
Comment (richardmin) [inline server/file_handler.cc:58]: Consider setting the content-length here? It's optional though.

File: server/file_handler_test.cc [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    EXPECT_EQ(ret, RequestHandler::OK);
+    std::string raw_req = "GET /static1/text_f.txt HTTP/1.1\r\nContent-Length: length\r\n\r\n";
+    auto req = Request::Parse(raw_req);
+
+    Response resp;
+    auto handler_status = handler.HandleRequest(*req,&resp);
+
+    EXPECT_EQ(handler_status,RequestHandler::FILE_NOT_FOUND);
+    EXPECT_EQ(resp.ToString(),"HTTP/1.1 404 Not Found\r\nContent-Type: text/html\r\n\r\n");
+}
```
Comment (richardmin) [inline server/file_handler_test.cc]: We should keep the new line character on these files!

File: server/httpRequest.cc [hunk 11]
```diff
@@ -1,10 +1,26 @@
...
+std::unique_ptr<Request> Request::Parse(const std::string& raw_request){
+	Request* req = new Request();
+
+	if(req->ParseRequest(raw_request)){
+		std::unique_ptr<Request> ret;
+		ret.reset(req);
+		return std::move(ret);
+	}
+	else{
+		return std::unique_ptr<Request>(nullptr);
```
Comment (richardmin) [inline server/httpRequest.cc]: Why are these unique_ptrs?

File: server/httpRequest.h [hunk 12]
```diff
@@ -3,30 +3,58 @@
...
+// the API when implementing the reverse proxy. Use your good judgment.
+
+// Represents an HTTP Request.
+//
+// Usage:
+//   auto request = Request::Parse(raw_request);
+class Request {
+ public:
+ 	//return nullptr if nnot successful
+ 	static std::unique_ptr<Request> Parse(const std::string& raw_request);
```
Comment (richardmin) [inline server/httpRequest.h:21]: Do these need to be unique_ptr?

File: server/httpResponse.h [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+  
+  void SetStatus(const ResponseCode response_code);
+  void AddHeader(const std::string& header_name, const std::string& header_value);
+  void SetBody(const std::string& body);
+  
+  //for the moment it only has HTTP/1.1
+  std::string ToString();
+
+private:
+	std::string reason_phrase_;
```
Comment (richardmin) [inline server/httpResponse.h:32]: Tab issue

**Review by richardmin — CHANGES_REQUESTED** [review]

File: server/httpRequest.cc (hunk 11 above)
Comment (ghost) [reply server/httpRequest.cc]: That's the Request interface, we have to use unique_ptr

File: server/file_handler.cc (hunk 7 above)
Comment (ghost) [reply server/file_handler.cc:32]: Ok I will leave it here for now just for safety

File: server/file_handler.cc (hunk 9 above)
Comment (hsjfelix) [inline server/file_handler.cc]: I think using this interface we have to set it with the AddHeader functions?

Comment (ghost) [reply server/file_handler.cc:58]: I think we have to use the AddHeader function for this?

File: server/file_handler.cc [hunk 14]
```diff
@@ -9,9 +9,9 @@ RequestHandler::Status StaticFileHandler::Init(const std::string& uri_prefix, co
 	//assume the config argument has something like
 	// root /path1
 	//check if such a line exists
-	auto root_config_tokens = config.find("root");
-	if(root_config_tokens.size() < 2){
-		printf("StaticfileHandler.Init: Invalid config");
+	std::vector<std::string> root_config_tokens = config.find("root");
+	if( root_config_tokens.size() < 2 ){
```
Comment (richardmin) [inline server/file_handler.cc]: what I meant was 
`if (root_config_tokens.size() < 2) {`
not 
`if( root_config_tokens.size() < 2  ){`

**Review by richardmin — CHANGES_REQUESTED** [review]

**Comment by richardmin** [comment]
Also, files_served.jpg isn't actually an image lol

File: server/file_handler.cc (hunk 14 above)
Comment (ghost) [reply server/file_handler.cc]: Anyway..

**Comment by ghost** [comment]
The jpg file isn't supposed to be a image file, I just use it as a to test http content type

File: server/file_handler.cc (hunk 9 above)
Comment (richardmin) [reply server/file_handler.cc:58]: Yes.

File: server/file_handler.cc (hunk 5 above)
Comment (richardmin) [reply server/file_handler.cc:21]: @hsjfelix can you fix this?

File: server/httpRequest.h (hunk 12 above)
Comment (richardmin) [reply server/httpRequest.h:21]: @hsjfelix

**Review by richardmin — CHANGES_REQUESTED** [review]
Please fix the whitespace (and othe rminor) issues

File: server/file_handler.cc (hunk 5 above)
Comment (ghost) [reply server/file_handler.cc:21]: what exactly is the issue?..

File: server/httpRequest.h (hunk 12 above)
Comment (ghost) [reply server/httpRequest.h:21]: Yeah we have to that's the interface

**Comment by ghost** [comment]
Ok that's it. Unless there are big issues I will just leave it like that.

**Review by richardmin — APPROVED** [review]
Looks good!

**Pull request merged**